cookie-manager.js

```diff
@@ -0,0 +1,228 @@
+                continue;
+            }
+
+            if (cm_user_preferences[cookie_category['category-name']] === 'off') {
+                console.info(`Cookie "${cookie_name}" is listed under category "${cookie_category['category-name']}"; user preferences opts out of this category; deleting.`);
+                deleteCookie(cookie_name);
+                continue;
+            }
+
+            console.info(`Cookie "${cookie_name}" is listed under category "${cookie_category['category-name']}"; user preferences opts in-to this category; cleared for use.`);
+        }
+
+        console.debug(`Finishing processing all cookies.`);
+    };
+
+    const getUserPreferences = function(cm_cookie) {
+
+        const cookie = getCookie(cm_cookie);
+
+        if (!cookie) {
+            return false;
+        }
+
+        try {
+            return JSON.parse(cookie);
+        } catch (e) {
+            console.error(`Unable to parse user preference cookie "${cm_cookie}" as JSON.`, e);
+            return false;
+        }
+    };
+
+    const getCookieCategoryFromManifest = function(cookie_name, configuration) {
+
+        const cookie_manifest = configuration['cookie-manifest'];
+
+        for (var i = 0; i < cookie_manifest.length; i++) {
+            const category_cookies = cookie_manifest[i]['cookies'];
+            for (var x = 0; x < category_cookies.length; x++) {
+                const cookie_prefix = category_cookies[x];
+                if (cookie_name.startsWith(cookie_prefix)) {
+                    console.debug(`Cookie "${cookie_name}" found in manifest.`);
+                    return cookie_manifest[i];
+                }
+            }
+        }
+        console.debug(`Cookie "${cookie_name}" NOT found in manifest.`);
+        return false;
+    };
+
+    const getCookie = function(cookie_name) {
+        const name = cookie_name + "=";
+        const decoded_cookie = decodeURIComponent(document.cookie);
+        const cookie_array = decoded_cookie.split(';');
+
+        for(var i = 0; i < cookie_array.length; i++) {
+            let cookie_part = cookie_array[i];
+            while (cookie_part.charAt(0) === ' ') {
+                cookie_part = cookie_part.substring(1);
+            }
+            if (cookie_part.indexOf(name) === 0) {
+                return cookie_part.substring(name.length, cookie_part.length);
+            }
+        }
+        return false;
+    };
+
+    const deleteCookie = function(cookie_name) {
+        document.cookie = cookie_name + '=; expires=Thu, 01 Jan 1970 00:00:01 GMT;';
+        console.debug(`Deleted cookie "${cookie_name}"`);
+    };
+
+    const setCookie = function(cookie_name, cookie_value, cookie_expiry_days) {
+        const date = new Date();
+        date.setTime(date.getTime() + (cookie_expiry_days * 24 * 60 * 60 * 1000));
+        const expires = "expires="+date.toUTCString();
+        document.cookie = cookie_name + "=" + encodeURIComponent(cookie_value) + ";" + expires + ";path=/";
```

Comment:
is there worth in setting the cookie as a secure cookie?  Often something commented on it ITHC's

Reply:
Yeah, I don't see why we cannot use the Secure attribute.

Reply:
After implementing this, it turns out we need to be careful. Running the scripts locally on a unsecured connection will impact testing. Therefore, I have now introduced a new property for the `cookie-manager.json` called `user-preference-cookie-secure`. If set, and set to `true` we append the `;secure` flag to the cookie. On production, and on test environments where we have **https** enabled, we can of-course set this.

Reply:
As a bonus, we have also added `user-preference-cookie-expiry-days` which allows us to override the expiry date of the cookie from the default of **365 days**